server/backend/solc.go

```diff
@@ -0,0 +1,172 @@
+package backend
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var versionRegexp = regexp.MustCompile(`([0-9]+)\.([0-9]+)\.([0-9]+)`)
+
+// Contract contains information about a compiled contract, alongside its code and runtime code.
+type Contract struct {
+	Code        string       `json:"code"`
+	RuntimeCode string       `json:"runtime-code"`
+	Info        ContractInfo `json:"info"`
+}
+
+// ContractInfo contains information about a compiled contract, including access
+// to the ABI definition, source mapping, user and developer docs, and metadata.
+//
+// Depending on the source, language version, compiler version, and compiler
+// options will provide information about how the contract was compiled.
+type ContractInfo struct {
+	Source          string      `json:"source"`
+	Language        string      `json:"language"`
+	LanguageVersion string      `json:"languageVersion"`
+	CompilerVersion string      `json:"compilerVersion"`
+	CompilerOptions string      `json:"compilerOptions"`
+	SrcMap          string      `json:"srcMap"`
+	SrcMapRuntime   string      `json:"srcMapRuntime"`
+	AbiDefinition   interface{} `json:"abiDefinition"`
+	UserDoc         interface{} `json:"userDoc"`
+	DeveloperDoc    interface{} `json:"developerDoc"`
+	Metadata        string      `json:"metadata"`
+}
+
+// Solidity contains information about the solidity compiler.
+type Solidity struct {
+	Path, Version       string
+	Major, Minor, Patch int
+}
+
+// --combined-output format
+type solcOutput struct {
+	Contracts map[string]struct {
+		BinRuntime                                  string `json:"bin-runtime"`
+		SrcMapRuntime                               string `json:"srcmap-runtime"`
+		Bin, SrcMap, Abi, Devdoc, Userdoc, Metadata string
+	}
+	Version string
+}
+
+func (s *Solidity) makeArgs() ([]string, error) {
+	dir, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+	return []string{
+		"run", "-i", "-v", dir + ":/workdir", "-w", "/workdir", "ethereum/solc:" + strconv.Itoa(s.Major) + "." + strconv.Itoa(s.Minor) + "." + strconv.Itoa(s.Patch),
+		"--combined-json",
+		"bin,bin-runtime,srcmap,srcmap-runtime,abi,userdoc,devdoc,metadata",
+		"--optimize", // code optimizer switched on
+	}, nil
+}
+
+// SolidityVersion scan the source code and parse the version from pragma attribute
+func SolidityVersion(source string) (*Solidity, error) {
+	var err error
+	for _, line := range strings.Split(source, "\n") {
+		if strings.HasPrefix(strings.TrimSpace(line), "pragma solidity") {
+			matches := versionRegexp.FindStringSubmatch(line)
```

Comment:
I think this is still too strict (e.g. won't `pragma solidity 5.0;` fail to parse?), and potentially incorrect - it still ignores the potential complexity of multiple matching rules (e.g. they might be telling us explicitly which version(s) _not_ to use).
Once we have `pragma solidity`, we could just take whatever is between that and the next `;`, and use it in full (`Major`, `Minor`, and `Patch` are only used to recreate the original anyways, right?). This way we would never misinterpret them, and would require a strict single version selection by the user, which is exactly the docker image tag to use.

Reply:
@jmank88 
> take whatever is between that and the next ;
not a good idea because this could lead to command injection -ie  `docker run -it ethereum/solc:fake_version || some bad command here to run || the rest of the parameters`

So I'd prefer to leave as it is, but I could scan for additional symbols (<>^~) and require to set exact version instead.

Reply:
Validation still certainly make sense.

So if someone deploys with e.g. `nightly-0.5.0-f6f0cecc2f64f4fcda3da467bbe306b5859729aa`, we are just not going to support that?

Reply:
`pragma solidity` is actually not even required.

What is the problem with using a separate input to select the version?

Reply:
IMHO we should reconsider a separate input. We can do a best-effort auto-select based on this method, and let them choose whatever they need if it doesn't work. We can easily improve on the remix selector, by allowing for simple searching/auto-complete.

Reply:
> Validation still certainly make sense.
> 
> So if someone deploys with e.g. `nightly-0.5.0-f6f0cecc2f64f4fcda3da467bbe306b5859729aa`, we are just not going to support that?

hmm I'd prefer not to support such cases, because it seems like there are differences between solc versions (https://github.com/ethereum/solidity/releases) and solc docker image versions (https://hub.docker.com/r/ethereum/solc/tags) so better to have have strict match

Reply:
> `pragma solidity` is actually not even required.
> 
> What is the problem with using a separate input to select the version?

that's possible as well but maybe be separate ticket?(also including parsing the sol code on a client side and preselecting the proper version from the list)

Reply:
I don't think that is a good reason to abandon a subset of use cases. We can do better.

Can we at least match the versions available via Remix?
![remix-solidity](https://user-images.githubusercontent.com/1194128/55098514-ec800580-508b-11e9-879d-6b45e67160dd.png)

Reply:
I believe they compile solc by themselves, we use docker image, so yes we could show the list of docker image tags but it wouldn't be pretty useful (ie there is no 0.6.0 version in docker image)

Reply:
I'd prefer user to use web3 to compile and deploy and then use explorer to verify (that why I vote for using the same approach in web3 and explorer)

Reply:
That is an unfortunate limitation, but why wouldn't it be useful? There are plenty of other intermediate versions that a user may need to produce the exact bytecode that they deployed. Why wouldn't we offer them that opportunity?

Reply:
We can't force them to use web3. We have to accommodate verifying deployed bytecode that was compiled by unknown manual processes or tools.

Reply:
In this case it makes sense to implement the same feature in web3 to keep the same UX

Reply:
@treeder do you want to allow user to select any solc version in both explorer and web3? or it should be both - trying to autoselect the solc version or user could force it?

Reply:
I'm all for aligning `web3` and explorer, but isn't the current approach broken - full stop?

A simple case: `pragma solidity <0.5.1;` will select `0.5.1` and fail.

Reply:
IMO both web3 and explorer can make a best-effort default/suggestion (the current approach is a start), but there has to be a way to override to a specific version.

Reply:
For `web3`: Even for the basic case of e.g. `pragma solidity >=1.0.0;`, do we really want to compile with `v1.0.0`? Or do we want the latest stable version (maybe `v5.x.y`) for what is likely a better result?

Reply:
> In this case it makes sense to implement the same feature in web3 to keep the same UX

I agree they should be somewhat aligned, but they are different use cases. `web3` should be trying to produce the best possible output (e.g. latest valid stable makes sense). Explorer verification is intended to reproduce a previous action exactly, and must be precise and customizable in order to do so.

Reply:
Yes, a user could modify the pragma to try to inform our decision (e.g. bump from `>=1.00` to `>=5.1.3`), but that is a misuse of what it is designed for. It expresses compatibility constraints.

Reply:
so let me then add a way to specify a version in both explorer and web3 and leave auto-suggestion(the dumb one we have now - not sure it's a good idea to try to resolve version constraints) if nothing was selected

Reply:
I agree that resolving complex constraints could be a mess, and can be improved later instead. As long as users can make an explicit custom choice, then they aren't left behind if the suggestion is incompatible or outdated.